### PR TITLE
Feat/disable profitability trigger

### PR DESCRIPTION
**Pull request by vic-en**
**Before submitting this PR, please make sure**:

- [ ] Your code builds clean without any errors or warnings
- [ ] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
- Volatility would now automatically fallback to user set spreads if volatility is zero.
- Fees in profitability calculation is not done more accurately.
- Trigger to remove position is now based on farthest inactive position. i.e Profitability as the basis for removing position is now removed.

### CODE REVIEW

**Comment by PtrckM** [comment]
After creating buy and sell position price range:
- unable to display `status` command
```
2021-07-10 04:50:58,947 - 32519 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: unsupported operand type(s) for +: 'int' and 'str'
Traceback (most recent call last):
  File "/home/ubuntu/uni_v3_ite-2/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/home/ubuntu/uni_v3_ite-2/hummingbot/client/command/status_command.py", line 135, in status_check_all
    self._notify(await self.strategy_status())
  File "/home/ubuntu/uni_v3_ite-2/hummingbot/client/command/status_command.py", line 77, in strategy_status
    st_status = await self.strategy.format_status()
  File "/home/ubuntu/uni_v3_ite-2/hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py", line 199, in format_status
    pos_profitability_df = self.profitability_df()
  File "/home/ubuntu/uni_v3_ite-2/hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py", line 165, in profitability_df
    profit_data = self.calculate_profitability(position)
  File "/home/ubuntu/uni_v3_ite-2/hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py", line 146, in calculate_profitability
    tx_fee = sum(position.tx_fees)
```

image

- the TokenID doesn't display on the right pane
  - even though its a successful creation https://etherscan.io/tx/0x091f5401e9e0ef8ab12e6a84c5012e80ebc9e01f77fb09b7d54adc6d4facd04b https://etherscan.io/tx/0xf6513ca04c5429bcbc556fa373e9f148cea7d434cd9545d8fde46ce9d7a49c6b
image
image

- operation error started to show as well `/eth/uniswap/v3/remove-position`
```
2021-07-10 04:50:59,957 - 32519 - hummingbot.connector.connector.uniswap.uniswap_connector - NETWORK - Error removing range position, token_id: None, hb_id: DAI-USDC-1625892619873558
Traceback (most recent call last):
  File "/home/ubuntu/uni_v3_ite-2/hummingbot/connector/connector/uniswap_v3/uniswap_v3_connector.py", line 448, in _remove_position
    result = await self._api_request("post",
  File "/home/ubuntu/uni_v3_ite-2/hummingbot/connector/connector/uniswap/uniswap_connector.py", line 591, in _api_request
    raise IOError(f"Error fetching data from {url}. HTTP status is {response.status}.{err_msg}")
OSError: Error fetching data from https://localhost:5123/eth/uniswap/v3/remove-position. HTTP status is 500. Message: Operation Error
2021-07-10 04:50:59,962 - 32519 - hummingbot.connector.connector.uniswap.uniswap_connector - WARNING - Error fetching data from https://localhost:5123/eth/uniswap/v3/remove-position. HTTP Status is 500. Message: Operation Error
2021-07-10 04:50:59,963 - 32519 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1625892659.0, "hb_id": "DAI-USDC-1625892619873558", "event_name": "RangePositionFailureEvent", "event_source": "uniswap_v3"}
```

When using amm_arb
- uniswap keep initializing for USDC-WETH even though I'm running balancer <-> binance with DAI-USDC <-> USDC-USDT
  - PS: no amm_arb uniswap yet created. Also showing on development/staging branch (does not happen before)

image

File: hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py [hunk 1]
```diff
@@ -351,8 +351,28 @@ def execute_proposal(self, proposal):
 
     def range_position_remover(self):
```
Comment (Nullably) [inline hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py:352]: This should be named with fn naming convention e.g. `remove_farthest_position`.

File: hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py [hunk 2]
```diff
@@ -351,8 +351,28 @@ def execute_proposal(self, proposal):
...
         """
-        This function removes  positions that are out of range and have attained min profitability.
+        This removes the farthest position.
+        """
+        inactive_sells = [position for position in self.active_positions if position.lower_price > self._last_price]
+        inactive_buys = [position for position in self.active_positions if position.upper_price < self._last_price]
+        farthest_position = None
+        if len(inactive_sells) > 1:
+            for position in inactive_sells:
+                farthest_position = position if farthest_position is None else position
```
Comment (Nullably) [inline hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py:361]: This `if` expression doesn't make sense as it always returns position regardless.

File: hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py [hunk 3]
```diff
@@ -351,8 +351,28 @@ def execute_proposal(self, proposal):
...
+                farthest_position = position if position.lower_price > farthest_position.lower_price else position
```
Comment (Nullably) [inline hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py:362]: Same with above.

File: hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py [hunk 4]
```diff
@@ -351,8 +351,28 @@ def execute_proposal(self, proposal):
...
+        elif len(inactive_buys) > 1:
+            for position in inactive_buys:
+                farthest_position = position if farthest_position is None else position
```
Comment (Nullably) [inline hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py:365]: same with above.

File: hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py [hunk 5]
```diff
@@ -351,8 +351,28 @@ def execute_proposal(self, proposal):
...
+                farthest_position = position if position.upper_price < farthest_position.upper_price else position
```
Comment (Nullably) [inline hummingbot/strategy/uniswap_v3_lp/uniswap_v3_lp.py:366]: same with above.

**Review by Nullably — CHANGES_REQUESTED** [review]

File: hummingbot/connector/connector/uniswap_v3/uniswap_v3_connector.py [hunk 6]
```diff
@@ -423,17 +427,18 @@ async def _add_position(self,
...
 
-    def remove_position(self, hb_id: str, token_id: str, reducePercent: Decimal = Decimal("100.0")):
-        safe_ensure_future(self._remove_position(hb_id, token_id, reducePercent))
+    def remove_position(self, hb_id: str, token_id: str, reducePercent: Decimal = Decimal("100.0"), fee_estimate: bool = False):
+        safe_ensure_future(self._remove_position(hb_id, token_id, reducePercent, fee_estimate))
         # get the inflight order that has this token_id
         return hb_id
 
-    async def _remove_position(self, hb_id: str, token_id: str, reducePercent: Decimal):
+    async def _remove_position(self, hb_id: str, token_id: str, reducePercent: Decimal, fee_estimate: bool):
```
[the remark this answers is not in the document]
Comment (vic-en) [reply hummingbot/connector/connector/uniswap_v3/uniswap_v3_connector.py:435]: yes. can you suggest a better name?

[the remark this answers is not in the document]
Comment (Nullably) [reply hummingbot/connector/connector/uniswap_v3/uniswap_v3_connector.py:434]: This function wraps around `remove-position` end point, the change is probably too much at this stage (as you will need to change Gateway as well). I am ok with the function name, but the parameter `fee_estimate` could be something like `to_estimate_fee_only` and good doc string would help explaining.

**Comment by PtrckM** [comment]
Test with `use_volatility` = `True` or `False` on mainnet
- no message or confirmation if order is completed or not - due insufficient fund (gas fee)
   - instead throws HTTP status is 500
- incomplete `status` command 
image

- bot accepts 0 value for `base_token_amount` 
image
image
image

- logs and configs
[07132021.zip](https://github.com/CoinAlpha/hummingbot/files/6805966/07132021.zip)

File: hummingbot/connector/connector/uniswap_v3/uniswap_v3_connector.py [hunk 7]
```diff
@@ -423,17 +427,17 @@ async def _add_position(self,
...
-        safe_ensure_future(self._remove_position(hb_id, token_id, reducePercent))
-        # get the inflight order that has this token_id
+    def remove_position(self, hb_id: str, token_id: str, reducePercent: Decimal = Decimal("100.0"), fee_estimate: bool = False):
+        safe_ensure_future(self._remove_position(hb_id, token_id, reducePercent, fee_estimate))
         return hb_id
 
-    async def _remove_position(self, hb_id: str, token_id: str, reducePercent: Decimal):
+    async def _remove_position(self, hb_id: str, token_id: str, reducePercent: Decimal, fee_estimate: bool):
         """
         Calls add position end point to create/increase a range position.
```
Comment (Nullably) [inline hummingbot/connector/connector/uniswap_v3/uniswap_v3_connector.py:436]:
```suggestion
        Calls remove position end point to remove/decrease a range position.
```

**Comment by vic-en** [comment]
@PtrckM I addressed the issue with some configs allowing a value of zero. Kindly confirm.

**Review by Nullably — APPROVED** [review]
LGTM

**Review by PtrckM — APPROVED** [review]
- Fix issue value 0 on `base_token_amount` :white_check_mark: 

- Re-tested and confirmed working by QA.

**Pull request merged**